File: vignettes/22_human_in_the_loop/csharp/22_human_in_the_loop.cs
// Human-in-the-Loop Workflows — C#
//
// This sample demonstrates human-in-the-loop patterns in workflows:
//   1. Using RequestPort to pause a workflow and request human input.
//   2. Resuming a workflow with SendResponseAsync.
//   3. A multi-step approval pipeline with writer, checker, and approver.
//
// A RequestPort acts as a workflow node that emits a RequestInfoEvent when
// it receives a message, pausing execution until an ExternalResponse is
// provided via SendResponseAsync.
//
// Prerequisites:
//   - Ollama running locally with qwen3:8b pulled
//   - dotnet restore

using Microsoft.Agents.AI;
using Microsoft.Agents.AI.Workflows;
using Microsoft.Extensions.AI;
using OllamaSharp;

var endpoint = Environment.GetEnvironmentVariable("OLLAMA_ENDPOINT")
    ?? "http://localhost:11434";
var modelName = Environment.GetEnvironmentVariable("OLLAMA_MODEL_NAME")
    ?? "qwen3:8b";

// ══════════════════════════════════════════════════════════════════════════ //
//  1. Simple Review Workflow with Human-in-the-Loop                        //
// ══════════════════════════════════════════════════════════════════════════ //

Console.WriteLine("=== Review Workflow with Human-in-the-Loop ===\n");

var chatClient = new OllamaApiClient(new Uri(endpoint), modelName);

// Writer agent generates a draft from the user's topic.
ChatClientAgent writer = new(chatClient,
    "You write a short paragraph about the given topic.",
    "Writer");

// RequestPort pauses the workflow and surfaces the draft to a human reviewer.
// It receives a ReviewSignal (the draft text) and expects a string response.
RequestPort reviewPort = RequestPort.Create<ReviewSignal, string>("Review");

// After the human responds, this executor decides the outcome.
ReviewResultExecutor reviewResult = new();

var reviewWorkflow = new WorkflowBuilder(writer)
    .AddEdge(writer, reviewPort)
    .AddEdge(reviewPort, reviewResult)
    .WithOutputFrom(reviewResult)
    .Build();

List<ChatMessage> input = [new ChatMessage(ChatRole.User, "Write about the Julia programming language")];

await using StreamingRun reviewRun = await InProcessExecution.RunStreamingAsync(
    reviewWorkflow, input);

await foreach (WorkflowEvent evt in reviewRun.WatchStreamAsync())
{
    switch (evt)
    {
        case RequestInfoEvent requestInfoEvt:
            Console.WriteLine($"Workflow paused — request ID: {requestInfoEvt.Request.RequestId}");
            if (requestInfoEvt.Request.TryGetDataAs<ReviewSignal>(out var reviewData))
            {
                Console.WriteLine($"  Draft preview: {reviewData.Draft[..Math.Min(100, reviewData.Draft.Length)]}...");
            }

            // Simulate human approval
            Console.WriteLine("  Human response: approve");
            ExternalResponse response = requestInfoEvt.Request.CreateResponse("approve");
            await reviewRun.SendResponseAsync(response);
            break;

        case WorkflowOutputEvent outputEvt:
            Console.WriteLine($"\nOutput: {outputEvt.Data}");
            goto done_review;
    }
}
done_review:

// ══════════════════════════════════════════════════════════════════════════ //
//  2. Multi-Step Approval Pipeline                                         //
// ══════════════════════════════════════════════════════════════════════════ //

Console.WriteLine("\n=== Multi-Step Approval Pipeline ===\n");

ChatClientAgent pipelineWriter = new(chatClient,
    "You write a one-paragraph article about the topic.",
    "Writer");

ChatClientAgent factChecker = new(chatClient,
    "You review the text for factual accuracy. Output PASS or list corrections.",
    "FactChecker");

// Approval gate: pauses for human sign-off after fact-checking.
RequestPort approvalPort = RequestPort.Create<ApprovalSignal, string>("Approval");

ApprovalResultExecutor approvalResult = new();

var pipeline = new WorkflowBuilder(pipelineWriter)
    .AddEdge(pipelineWriter, factChecker)
    .AddEdge(factChecker, approvalPort)
    .AddEdge(approvalPort, approvalResult)
    .WithOutputFrom(approvalResult)
    .Build();

List<ChatMessage> pipelineInput = [new ChatMessage(ChatRole.User,
    "The history of the Julia programming language")];

await using StreamingRun pipelineRun = await InProcessExecution.RunStreamingAsync(
    pipeline, pipelineInput);

await foreach (WorkflowEvent evt in pipelineRun.WatchStreamAsync())
{
    switch (evt)
    {
        case RequestInfoEvent requestInfoEvt:
            Console.WriteLine($"Approval requested — request ID: {requestInfoEvt.Request.RequestId}");
            if (requestInfoEvt.Request.TryGetDataAs<ApprovalSignal>(out var approvalData))
            {
                Console.WriteLine($"  Content preview: {approvalData.Content[..Math.Min(200, approvalData.Content.Length)]}...");
            }

            // Human approves
            Console.WriteLine("  Human response: approve");
            ExternalResponse approvalResponse = requestInfoEvt.Request.CreateResponse("approve");
            await pipelineRun.SendResponseAsync(approvalResponse);
            break;

        case WorkflowOutputEvent outputEvt:
            Console.WriteLine($"\n{outputEvt.Data}");
            goto done_pipeline;
    }
}
done_pipeline:
Console.WriteLine();

// ══════════════════════════════════════════════════════════════════════════ //
//  Signal & Executor Definitions                                            //
// ══════════════════════════════════════════════════════════════════════════ //

/// <summary>Signal sent to the review RequestPort carrying the draft text.</summary>
internal sealed record ReviewSignal(string Draft);

/// <summary>Signal sent to the approval RequestPort carrying fact-checked content.</summary>
internal sealed record ApprovalSignal(string Content);

/// <summary>
/// Processes the human reviewer's response (the string returned through the
/// RequestPort) and yields a final verdict as workflow output.
/// </summary>
[YieldsOutput(typeof(string))]
internal sealed class ReviewResultExecutor() : Executor<string>("ReviewResult")
{
    public override async ValueTask HandleAsync(
        string message,
        IWorkflowContext context,
        CancellationToken cancellationToken = default)
    {
        if (message.Trim().Equals("approve", StringComparison.OrdinalIgnoreCase))
        {
            await context.YieldOutputAsync("✅ Approved and published", cancellationToken);
        }
        else
        {
            await context.YieldOutputAsync($"❌ Revision needed: {message}", cancellationToken);
        }
    }
}

/// <summary>
/// Processes the human approver's response and yields a publication decision.
/// </summary>
[YieldsOutput(typeof(string))]
internal sealed class ApprovalResultExecutor() : Executor<string>("ApprovalResult")
{
    public override async ValueTask HandleAsync(
        string message,
        IWorkflowContext context,
        CancellationToken cancellationToken = default)
    {
        if (message.Trim().Equals("approve", StringComparison.OrdinalIgnoreCase))
        {
            await context.YieldOutputAsync("✅ Published", cancellationToken);
        }
        else
        {
            await context.YieldOutputAsync($"❌ Rejected: {message}", cancellationToken);
        }
    }
}

File: vignettes/22_human_in_the_loop/python/22_human_in_the_loop.py
"""
Human-in-the-Loop Workflows — Python

This sample demonstrates human-in-the-loop patterns in workflows:
  1. Using request_info to pause a workflow and request human input.
  2. Resuming a workflow with responses.
  3. Function approval requests for sensitive tool calls.
  4. A multi-step approval pipeline with writer, checker, and approver.

Prerequisites:
  - Ollama running locally with `qwen3:8b` pulled
  - pip install agent-framework-ollama
"""

import asyncio

from agent_framework import (
    AgentExecutor,
    AgentExecutorRequest,
    AgentExecutorResponse,
    Content,
    Executor,
    Message,
    WorkflowBuilder,
    WorkflowContext,
    executor,
    handler,
    response_handler,
)
from agent_framework.ollama import OllamaChatClient
from typing_extensions import Never


# ── Review Executor with request_info ────────────────────────────────────────


class DraftReviewRequest:
    """Data class for draft review requests."""

    def __init__(self, prompt: str, draft: str):
        self.prompt = prompt
        self.draft = draft


class ReviewExecutor(Executor):
    """Pauses the workflow to request human feedback on a draft."""

    def __init__(self) -> None:
        super().__init__(id="reviewer")

    @handler
    async def on_draft(
        self,
        draft: AgentExecutorResponse,
        ctx: WorkflowContext[Never, str],
    ) -> None:
        text = draft.agent_response.text or ""
        await ctx.request_info(
            request_data=DraftReviewRequest(
                prompt=f"Please review this draft: {text[:100]}...",
                draft=text,
            ),
            response_type=str,
        )

    @response_handler
    async def on_human_feedback(
        self,
        original_request: DraftReviewRequest,
        feedback: str,
        ctx: WorkflowContext[Never, str],
    ) -> None:
        """Process human feedback after workflow resumes."""
        if feedback.strip().lower() == "approve":
            await ctx.yield_output(original_request.draft)
        else:
            await ctx.yield_output(f"Revision needed: {feedback}")


# ── Approval Gate Executor ───────────────────────────────────────────────────


class ApprovalRequest:
    """Data class for approval gate requests."""

    def __init__(self, prompt: str, content: str):
        self.prompt = prompt
        self.content = content


class ApprovalGate(Executor):
    """Requests final human approval before publishing."""

    def __init__(self) -> None:
        super().__init__(id="approver")

    @handler
    async def on_content(
        self,
        content: AgentExecutorResponse,
        ctx: WorkflowContext[Never, str],
    ) -> None:
        text = content.agent_response.text or ""
        await ctx.request_info(
            request_data=ApprovalRequest(
                prompt=f"Fact-check result: {text[:200]}\n\nApprove for publication?",
                content=text,
            ),
            response_type=str,
        )

    @response_handler
    async def on_approval(
        self,
        original_request: ApprovalRequest,
        feedback: str,
        ctx: WorkflowContext[Never, str],
    ) -> None:
        if feedback.strip().lower() == "approve":
            await ctx.yield_output(f"✅ Published: {original_request.content}")
        else:
            await ctx.yield_output(f"❌ Rejected: {feedback}")


# ── Main ─────────────────────────────────────────────────────────────────────


async def main() -> None:
    client = OllamaChatClient(
        host="http://localhost:11434",
        model="qwen3:8b",
    )

    # ── 1. Simple review workflow ────────────────────────────────────────
    print("=== Review Workflow with Human-in-the-Loop ===\n")

    writer = AgentExecutor(
        client.as_agent(
            name="Writer",
            instructions="Write a short paragraph about the given topic.",
        )
    )
    reviewer = ReviewExecutor()

    review_workflow = (
        WorkflowBuilder(start_executor=writer)
        .add_edge(writer, reviewer)
        .build()
    )

    # Step 1: Run until human input is needed
    events = await review_workflow.run("Write about the Julia programming language")
    request_info_events = events.get_request_info_events()
    print(f"State: {events.get_final_state()}")
    print(f"Pending requests: {len(request_info_events)}")

    # Step 2: Provide human feedback and resume
    if request_info_events:
        request_id = request_info_events[0].request_id
        print(f"  Request data: {request_info_events[0].data.prompt}")

        # Simulate human approving the draft
        events = await review_workflow.run(
            responses={request_id: "approve"}
        )
        outputs = events.get_outputs()
        print(f"\nApproved output:\n  {outputs[0][:200]}...")

    # ── 2. Multi-step approval pipeline ──────────────────────────────────
    print("\n\n=== Multi-Step Approval Pipeline ===\n")

    pipeline_writer = AgentExecutor(
        client.as_agent(
            name="Writer",
            instructions="Write a one-paragraph article about the topic.",
        )
    )
    fact_checker = AgentExecutor(
        client.as_agent(
            name="FactChecker",
            instructions="Review the text for factual accuracy. Output PASS or list corrections.",
        )
    )
    approver = ApprovalGate()

    pipeline = (
        WorkflowBuilder(start_executor=pipeline_writer)
        .add_edge(pipeline_writer, fact_checker)
        .add_edge(fact_checker, approver)
        .build()
    )

    # Run until approval is needed
    events = await pipeline.run("The history of the Julia programming language")
    request_info_events = events.get_request_info_events()
    print(f"State: {events.get_final_state()}")

    if request_info_events:
        req = request_info_events[0]
        print(f"Approval requested:\n  {req.data.prompt[:200]}...")

        # Human approves
        events = await pipeline.run(responses={req.request_id: "approve"})
        outputs = events.get_outputs()
        for output in outputs:
            print(f"\n{output}")

    # ── 3. Function approval request (conceptual) ──────────────────────
    print("\n\n=== Function Approval Request (Conceptual) ===\n")

    agent = client.as_agent(
        name="FileManager",
        instructions="You manage files. Describe what you would do when asked to delete a file.",
    )

    # Without tools requiring approval defined, the agent just responds with text.
    # In a real scenario, tools with requires_approval=True would trigger
    # function_approval_request content that the caller must approve/reject.
    response = await agent.run("Please delete old_report.txt")
    print(f"Agent response: {response.text[:300]}")
    print("\n(Full function-approval requires tools with requires_approval=True)")


if __name__ == "__main__":
    asyncio.run(main())
